lms/djangoapps/completion/models.py

```diff
@@ -0,0 +1,72 @@
+"""
+Completion tracking and aggregation models.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from django.contrib.auth.models import User
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import ugettext as _
+from model_utils.models import TimeStampedModel
+
+from openedx.core.djangoapps.xmodule_django.models import CourseKeyField, UsageKeyField
+
+
+def validate_ratio(value):
+    """
+    Verify that the passed value is between 0.0 and 1.0.
+    """
+    if not 0.0 <= value <= 1.0:
+        raise ValidationError(_('{value} must be between 0.0 and 1.0').format(value=value))
+
+
+class BlockCompletionManager(models.Manager):
+    """
+    Custom manager for BlockCompletion model.
+
+    Adds update_or_create method.
+    """
+
+    def update_or_create(self, user, course_key, block_key, completed):
```

Comment:
[`update_or_create ` is actually a standard method on the default Manager already](https://docs.djangoproject.com/en/1.8/ref/models/querysets/#update-or-create), and this is shadowing that base class method. Can you just use that instead?

Reply:
This has two customizations over the standard create_or_update, one of which is important, and the other which is just a convenience: 

Important: It only updates the entry when the existing entry has changed its completion value.  This prevents the modified timestamp from getting updated when it shouldn't, which will prove important for determining when to perform a recalculation of aggregators.

Convenience: The caller is not required to pass a block type, because it can be inferred from the block key.

I agree it's not great to shadow create_or_update.  I could go one of two ways with this:

1. Eliminate the convenience feature, so this has the same behavior as the super() version, but still performs the check that values have changed before updating.
2. Rename this (maybe `submit_completion`) and keep the convenience function.  I'm tempted to go with 2, to make it clear that we're working with a custom API.

Reply:
Got it. Thanks for renaming - I think it's better not to shadow methods and change the behavior :)